File: tofurengo_js/src/tag_parser.js
/**
 * Tag parsing, replacement protocols, and placeholder escaping utilities for client-side JavaScript.
 *
 * Provides core parsing logic, token escaping, issue collection data structures,
 * and high-level string transformation pipeline functionality.
 */

// Control character used as a temporary placeholder for escaped opening braces '{{'
export const MARK_LB = "\u0002";

// Pattern string for matching a single tag enclosed in single braces: {glyph_name key=value ...}
// Disallows newlines inside tags and captures raw content inside group 1.
// Note: Defined as a plain string to avoid shared mutable RegExp state (lastIndex).
export const TAG_PATTERN_SOURCE =
  "\\{([ \\t]*[A-Za-z0-9_\\-]+(?:[ \\t]+[^}\\r\\n]+)?)\\}";

/**
 * Severity levels for tag processing diagnostics.
 *
 * @readonly
 * @enum {string}
 */
export const IssueLevel = {
  WARNING: "warning",
  ERROR: "error",
};

/**
 * Represents a warning or error encountered during tag processing.
 */
export class TagIssue {
  /**
   * @param {string} code - Machine-readable issue category identifier.
   * @param {string} message - Human-readable failure explanation.
   * @param {string} [level=IssueLevel.ERROR] - Severity level of the issue.
   * @param {Record<string, any>} [details={}] - Additional contextual metadata.
   */
  constructor(code, message, level = IssueLevel.ERROR, details = {}) {
    /** @type {string} */
    this.code = code;

    /** @type {string} */
    this.message = message;

    /** @type {string} */
    this.level = level;

    /** @type {Record<string, any>} */
    this.details = details;
  }
}

/**
 * Structured representation of a parsed glyph tag's contents.
 */
export class ParsedTag {
  /**
   * @param {string} glyphName - Primary glyph identifier.
   * @param {Record<string, string>} [properties={}] - Extracted key-value pairs.
   * @param {string} [rawContent=""] - Original unparsed tag content inside braces.
   */
  constructor(glyphName, properties = {}, rawContent = "") {
    /** @type {string} */
    this.glyphName = glyphName;

    /** @type {Record<string, string>} */
    this.properties = properties;

    /** @type {string} */
    this.rawContent = rawContent;

    // Shortcut accessors
    /** @type {string|null} */
    this.b = properties["b"] || null;

    /** @type {string|null} */
    this.v = properties["v"] || null;

    /** @type {string|null} */
    this.set = properties["set"] || null;
  }

  /**
   * Construct a ParsedTag instance directly from raw tag inner content.
   *
   * @param {string} content - Raw inner string extracted from inside a tag.
   * @returns {ParsedTag} Newly created ParsedTag instance.
   */
  static fromContent(content) {
    const rawStr = content.trim();
    const tokens = rawStr.split(/[ \t]+/);

    if (tokens.length === 0 || tokens[0] === "") {
      return new ParsedTag("", {}, rawStr);
    }

    let glyphName = "";
    let attrTokens = [];

    // Check if the first token is a key-value pair (missing glyph_name)
    if (tokens[0].includes("=")) {
      glyphName = "";
      attrTokens = tokens;
    } else {
      glyphName = tokens[0];
      attrTokens = tokens.slice(1);
    }

    const properties = {};
    let lastKey = null;

    for (const token of attrTokens) {
      if (token.includes("=")) {
        const eqIndex = token.indexOf("=");
        const key = token.slice(0, eqIndex);
        const value = token.slice(eqIndex + 1);
        properties[key] = value;
        lastKey = key;
      } else if (lastKey !== null) {
        // Append multi-token values (e.g., space-separated UCS sequences like 'U+845B U+E0103')
        properties[lastKey] = `${properties[lastKey]} ${token}`;
      }
    }

    return new ParsedTag(glyphName, properties, rawStr);
  }
}

/**
 * Utility class for handling brace escaping, restoration, and tag parsing pipelines.
 */
export class TagParser {
  /**
   * Replace escaped double-brace sequences with a temporary control character placeholder.
   *
   * @param {string} text - Input text containing double braces `{{`.
   * @returns {string} Text with double braces replaced by placeholder tokens.
   */
  static escapeTokens(text) {
    return text.replace(/\{\{/g, MARK_LB);
  }

  /**
   * Restore internal placeholders back to double-brace escape sequences (`{{`).
   *
   * @param {string} text - Text containing temporary placeholder tokens.
   * @returns {string} Text with placeholders restored to literal `{{`.
   */
  static restoreTokensPreserveEscape(text) {
    return text.replace(new RegExp(MARK_LB, "g"), "{{");
  }

  /**
   * Restore internal placeholders to single unescaped braces (`{`).
   *
   * @param {string} text - Text containing temporary placeholder tokens.
   * @returns {string} Text with placeholders converted to single `{`.
   */
  static restoreTokensUnescape(text) {
    return text.replace(new RegExp(MARK_LB, "g"), "{");
  }

  /**
   * Execute the full transformation pipeline: escape -> substitute -> restore.
   *
   * @param {string} text - Target text to process.
   * @param {function(ParsedTag, Array<TagIssue>): string} replacer - Callback receiving parsed tag and issues array.
   * @param {boolean} [unescape=true] - Convert preserved placeholders to single '{' if true.
   * @param {Array<TagIssue>|null} [issues=null] - Optional mutable array to collect encountered issues.
   * @returns {string} Transformed output text.
   */
  static processPipeline(text, replacer, unescape = true, issues = null) {
    const issueList = issues !== null ? issues : [];
    const escaped = this.escapeTokens(text);

    // Instantiate a fresh RegExp locally with global flag 'g' using the pattern string
    const globalPattern = new RegExp(TAG_PATTERN_SOURCE, "g");
    const substituted = escaped.replace(globalPattern, (match, content) => {
      const tag = ParsedTag.fromContent(content);
      return replacer(tag, issueList);
    });

    if (unescape) {
      return this.restoreTokensUnescape(substituted);
    }
    return this.restoreTokensPreserveEscape(substituted);
  }
}

File: tofurengo/src/tofurengo/tag_parser.py
"""
Tag parsing, replacement protocols, and placeholder escaping utilities.

This module provides data models and utilities for tag-based string processing,
including `TagIssue` for capturing validation warnings and errors, `ParsedTag` for
structured tag attribute access, `ReplaceFn` for substitution callbacks,
and `TagParser` for managing double-brace escaping.
"""

from dataclasses import dataclass, field
from enum import Enum
import re
from typing import Any, Protocol

# Control character used as a placeholder for escaped opening braces '{{'
MARK_LB: str = "\u0002"

# Matches a single tag enclosed in single braces: {glyph_name key=value ...}
# Allows optional leading horizontal whitespace before the glyph name,
# requires a valid glyph name identifier, disallows newlines, and captures
# the raw tag body inside the "content" named group.
TAG_PATTERN: re.Pattern[str] = re.compile(
    r"\{"
    r"(?P<content>[ \t]*[A-Za-z0-9_\-]+(?:[ \t]+[^}\r\n]+)?)"
    r"\}",
)


class IssueLevel(str, Enum):
    """
    Severity levels for tag processing diagnostics.
    """

    WARNING = "warning"
    ERROR = "error"


@dataclass
class TagIssue:
    """
    Represents a warning or error encountered during tag processing.

    Attributes:
        code (str): Machine-readable issue category identifier.
        message (str): Human-readable message explaining failure details.
        level (IssueLevel): Severity level (`IssueLevel.WARNING` or `IssueLevel.ERROR`).
        details (dict[str, Any]): Additional contextual metadata regarding the issue.
    """

    code: str
    message: str
    level: IssueLevel = IssueLevel.ERROR
    details: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """
        Convert the issue instance into a JSON-serializable dictionary.

        Returns:
            dict[str, Any]: Dictionary representation of the tag issue.
        """
        return {
            "code": self.code,
            "message": self.message,
            "level": self.level.value,
            "details": self.details,
        }


@dataclass
class ParsedTag:
    """
    Structured representation of a parsed glyph tag's contents.

    Attributes:
        glyph_name (str): The primary glyph identifier (e.g., 'MJ000001').
        b (str | None): Value of the 'b' (base UCS) property if present.
        v (str | None): Value of the 'v' (variation) property if present.
        set (str | None): Value of the 'set' property if present.
        properties (dict[str, str]): All extracted key-value pairs.
        raw_content (str): The original unparsed tag inner content string.
    """

    glyph_name: str
    b: str | None = None
    v: str | None = None
    set: str | None = None
    properties: dict[str, str] = field(default_factory=dict)
    raw_content: str = ""

    @classmethod
    def from_content(cls, content: str) -> "ParsedTag":
        """
        Construct a `ParsedTag` instance directly from raw tag inner content.

        Args:
            content (str): Raw string extracted from inside a tag.

        Returns:
            ParsedTag: Structured object with shortcut accessors for b, v, and set.
        """
        raw_str = content.strip()
        tokens = raw_str.split()
        if not tokens:
            return cls(glyph_name="", raw_content=raw_str)

        # Check if the first token is actually a key-value pair (missing glyph_name)
        if "=" in tokens[0]:
            glyph_name = ""
            attr_tokens = tokens
        else:
            glyph_name = tokens[0]
            attr_tokens = tokens[1:]

        properties: dict[str, str] = {}
        last_key: str | None = None

        for token in attr_tokens:
            if "=" in token:
                key, value = token.split("=", 1)
                properties[key] = value
                last_key = key
            elif last_key is not None:
                # Append multi-token values (e.g., space-separated UCS sequences like 'U+845B U+E0103')
                properties[last_key] = f"{properties[last_key]} {token}"

        return cls(
            glyph_name=glyph_name,
            b=properties.get("b"),
            v=properties.get("v"),
            set=properties.get("set"),
            properties=properties,
            raw_content=raw_str,
        )


class ReplaceFn(Protocol):
    """
    Protocol definition for tag match-replacement closures.

    Implementations are callable objects that accept a structured `ParsedTag`
    and a mutable issue list, returning a replacement string while appending any
    encountered issues.
    """

    def __call__(self, tag: ParsedTag, issues: list[TagIssue]) -> str:
        """
        Process a parsed tag object and record any non-fatal processing issues.

        Args:
            tag (ParsedTag): Structured representation of the tag to process.
            issues (list[TagIssue]): Mutable list to collect encountered issues.

        Returns:
            str: The replacement string to substitute into target text.
        """
        ...


class TagParser:
    """
    Utility class for handling brace escaping, restoration, and tag parsing.

    Provides mechanisms to temporarily protect escaped braces (`{{`), parse
    and substitute tag matches, and restore the preserved sequences into either
    unescaped (`{`) or original escaped (`{{`) representations.
    """

    @staticmethod
    def escape_tokens(text: str) -> str:
        """
        Replace escaped double-brace sequences with a temporary control character.

        Args:
            text (str): Input text containing potential `{{` escape sequences.

        Returns:
            str: Text with `{{` replaced by the internal placeholder character.
        """
        return text.replace("{{", MARK_LB)

    @staticmethod
    def restore_tokens_preserve_escape(text: str) -> str:
        """
        Restore internal placeholders back to double-brace escape sequences (`{{`).

        Args:
            text (str): Processed text containing placeholder characters.

        Returns:
            str: Text with placeholders restored to `{{`.
        """
        return text.replace(MARK_LB, "{{")

    @staticmethod
    def restore_tokens_unescape(text: str) -> str:
        """
        Restore internal placeholders to single unescaped braces (`{`).

        Args:
            text (str): Processed text containing placeholder characters.

        Returns:
            str: Text with placeholders converted to `{`.
        """
        return text.replace(MARK_LB, "{")

    @classmethod
    def process_pipeline(
        cls,
        text: str,
        replacer: ReplaceFn,
        unescape: bool = True,
        issues: list[TagIssue] | None = None,
    ) -> str:
        """
        Execute the full transformation pipeline: escape -> substitute -> restore.

        Args:
            text (str): Target text to process.
            replacer (ReplaceFn): Replacement callback implementing `ReplaceFn`.
            unescape (bool): If `True`, converts preserved placeholders to `{`.
                If `False`, preserves them as `{{`. Defaults to `True`.
            issues (list[TagIssue] | None): Optional mutable list to collect issues
                encountered during substitution. If `None`, an internal list is used.

        Returns:
            str: Transformed output text after tag substitution and brace restoration.
        """
        issue_list = issues if issues is not None else []
        escaped = cls.escape_tokens(text)

        def sub_callback(match: re.Match[str]) -> str:
            content = match.group("content")
            tag = ParsedTag.from_content(content)
            return replacer(tag, issue_list)

        substituted = TAG_PATTERN.sub(sub_callback, escaped)

        if unescape:
            return cls.restore_tokens_unescape(substituted)
        return cls.restore_tokens_preserve_escape(substituted)

